flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxProcessor.java

```diff
@@ -0,0 +1,319 @@
+
+		final Mailbox localMailbox = mailbox;
+
+		assert localMailbox.getState() == Mailbox.State.OPEN : "Mailbox must be opened!";
+
+		final MailboxDefaultActionContext defaultActionContext = new MailboxDefaultActionContext(this);
+
+		while (processMail(localMailbox)) {
+			mailboxDefaultAction.runDefaultAction(defaultActionContext);
+		}
+	}
+
+	/**
+	 * This method exists to run the current sources with the mailbox model in a compatibility mode. Once we drop the
+	 * old source interface for the new one (FLIP-27) this method can eventually go away.
+	 *
+	 * @param checkpointLock the task's checkpointing lock.
+	 * @throws MailboxStateException if mailbox is closed.
+	 * @throws InterruptedException on interruption.
+	 */
+	public void switchToLegacySourceCompatibilityMailboxLoop(
+		final Object checkpointLock) throws MailboxStateException, InterruptedException {
+
+		assert taskMailboxExecutor.isMailboxThread() :
+			"Legacy source compatibility mailbox loop must run in mailbox thread!";
+
+		while (isMailboxLoopRunning()) {
+
+			Runnable letter = mailbox.takeMail();
+
+			synchronized (checkpointLock) {
+				letter.run();
+			}
+		}
+	}
+
+	/**
+	 * Cancels the mailbox loop execution. All pending mailbox actions will not be executed anymore, if they are
+	 * instance of {@link java.util.concurrent.RunnableFuture}, they will be cancelled.
+	 */
+	public void cancelMailboxExecution() {
+		try {
+			List<Runnable> droppedRunnables = mailbox.clearAndPut(mailboxPoisonLetter);
+			FutureUtils.cancelRunnableFutures(droppedRunnables);
+		} catch (MailboxStateException msex) {
+			LOG.debug("Mailbox already closed in cancel().", msex);
+		}
+	}
+
+	/**
+	 * This method must be called to end the stream task when all actions for the tasks have been performed.
+	 */
+	public void allActionsCompleted() {
+		try {
+			if (taskMailboxExecutor.isMailboxThread()) {
+				mailboxLoopRunning = false;
+				ensureControlFlowSignalCheck();
+			} else {
+				mailbox.putFirst(mailboxPoisonLetter);
+			}
+		} catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+		} catch (MailboxStateException me) {
+			LOG.debug("Action context could not submit poison letter to mailbox.", me);
+		}
+	}
+
+	/**
+	 * This helper method handles all special actions from the mailbox. It returns true if the mailbox loop should
+	 * continue running, false if it should stop. In the current design, this method also evaluates all control flag
+	 * changes. This keeps the hot path in {@link #runMailboxLoop()} free from any other flag checking, at the cost
+	 * that all flag changes must make sure that the mailbox signals mailbox#hasMail.
+	 */
+	private boolean processMail(Mailbox mailbox) throws MailboxStateException, InterruptedException {
+
+		if (!mailbox.hasMail()) {
+			// We can directly return true because all changes to #isMailboxLoopRunning must be connected to
+			// mailbox.hasMail() == true.
+			return true;
```

Comment:
Why not remove this `if` at all here? Currently I see some workarounds that check that the mailbox is not empty, but making it all fragile and not easy to reason about, imo.

Reply:
Because it saves you to read a field in every loop iteration.

Reply:
If it's about optimization, what do you think about the following variant:
```java
if (mailbox.hasMail()) {
	// TODO consider batched draining into list and/or limit number of executed letters
	// Take letters in a non-blockingly and execute them.
	Optional<Runnable> maybeLetter;
	while (isMailboxLoopRunning() && (maybeLetter = mailbox.tryTakeMail()).isPresent()) {
		maybeLetter.get().run();
	}
}

// If the default action is currently not available, we can run a blocking mailbox execution until the default
// action becomes available again.
while (isDefaultActionUnavailable() && isMailboxLoopRunning()) {
	mailbox.takeMail().run();
}

return isMailboxLoopRunning();
```
?

Reply:
I was thinking about that, still this one reads two more fields per iteration than the other variant, at the cost of having to do a bit more in the uncommon path (adding a letter).

Reply:
The suggestion would also be only a partial fix, not covering the case where we want to terminate the loop.